google/cloud/storage/internal/curl_client_test.cc

```diff
@@ -46,306 +49,358 @@ class FailingCredentials : public Credentials {
   }
 };
 
-class CurlClientTest : public ::testing::Test {
+class CurlClientTest : public ::testing::Test,
+                       public ::testing::WithParamInterface<std::string> {
  protected:
-  static void SetUpTestCase() {
-    client_ = CurlClient::Create(std::make_shared<FailingCredentials>());
+  CurlClientTest() : endpoint_("CLOUD_STORAGE_TESTBENCH_ENDPOINT") {}
+
+  void SetUp() override {
+    std::string const error_type = GetParam();
+    if (error_type == "credentials-failure") {
+      client_ = CurlClient::Create(std::make_shared<FailingCredentials>());
+      expected_status_code_ = STATUS_ERROR_CODE;
+      expected_status_substr_ = STATUS_ERROR_MSG;
+    } else {
+      google::cloud::internal::SetEnv(
+          "CLOUD_STORAGE_TESTBENCH_ENDPOINT", "http://localhost:0");
+      client_ = CurlClient::Create(
+          ClientOptions(oauth2::CreateAnonymousCredentials())
+              .set_endpoint("http://localhost:0"));
+      expected_status_code_ = StatusCode::UNKNOWN;
+      expected_status_substr_ = "CURL error";
+    }
   }
 
-  static void TearDownTestCase() { client_.reset(); }
+  void TearDown() override { client_.reset(); }
 
-  static std::shared_ptr<CurlClient> client_;
-};
-
-std::shared_ptr<CurlClient> CurlClientTest::client_;
+  void TestCorrectFailureStatus(Status const& status) {
+    EXPECT_EQ(expected_status_code_, status.status_code());
+    EXPECT_THAT(status.error_message(), HasSubstr(expected_status_substr_));
+  }
 
-void TestCorrectFailureStatus(Status const& status) {
-  EXPECT_EQ(status.status_code(), STATUS_ERROR_CODE);
-  EXPECT_EQ(status.error_message(), STATUS_ERROR_MSG);
-}
+  std::shared_ptr<CurlClient> client_;
+  long expected_status_code_;
+  std::string expected_status_substr_;
+  testing_util::EnvironmentVariableRestore endpoint_;
```

Comment:
endpoint_.TearDown() in (a new) ~CurlClientTest()?

Reply:
Indeed, for some reason I thought `endpoint_` had a destructor that did that, thanks for checking.